### PR TITLE
[#9536] Instructor Courses Page E2E Test

### PR DESCRIPTION
**Pull request by jtankw3**
Part of #9536

**Outline of Solution**
- Add UserId parsing to Actions with Backdoor authentication to access data as user.
- Add getArchivedCourse to Backdoor to verify archived course status.
- Add InstructorCoursesPage to represent instructor courses page.
- Add InstructorCoursesPageE2ETest which tests for 
    - Verify loaded data
    - Verify statistics
    - Verify cannot modify without permissions
    - Add new course
    - Archive course
    - Unarchive course
    - Move active course to recycle bin
    - Restore active course
    - Move archived course to recycle bin
    - Restore archived course
    - Permanently delete course
    - Restore all
    - Permanently delete all

### CODE REVIEW

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        coursesPage.verifyStatusMessage("All courses have been permanently deleted.");
+        coursesPage.verifyNumActiveCourses(1);
+        coursesPage.verifyNumArchivedCourses(0);
+        coursesPage.verifyNumDeletedCourses(0);
+        verifyAbsentInDatastore(courses[1]);
+        verifyAbsentInDatastore(courses[2]);
+
+    }
+
+    private void verifyCourseArchivedInDatastore(String instructorId, CourseAttributes course) {
```
Comment (rrtheonlyone) [inline src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java:180]: Why are these methods here instead of in backdoor? (just like `verifyPresentInDatastore` for example)

File: src/e2e/java/teammates/e2e/util/BackDoor.java [hunk 2]
```diff
@@ -282,6 +283,47 @@ public static CourseData getCourseData(String courseId) {
...
+        return CourseAttributes.builder(courseData.getCourseId())
+                .withName(courseData.getCourseName())
+                .withTimezone(ZoneId.of(courseData.getTimeZone()))
+                .build();
+    }
+
+    /**
+     * Gets archived course data from the datastore.
+     */
+    public static CourseData getArchivedCourseData(String instructorId, String courseId) {
```
Comment (rrtheonlyone) [inline src/e2e/java/teammates/e2e/util/BackDoor.java:298]: Is this method really needed? It looks like it is hitting the same endpoint as `getCourse`.

File: src/main/java/teammates/ui/webapi/action/Action.java [hunk 3]
```diff
@@ -97,6 +97,7 @@ public void checkAccessControl() {
     private void initAuthInfo() {
         if (Config.BACKDOOR_KEY.equals(req.getHeader("Backdoor-Key"))) {
             authType = AuthType.ALL_ACCESS;
+            userInfo = new UserInfo(getRequestParamValue(Const.ParamsNames.USER_ID));
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/action/Action.java:100]: Why was this added?

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        CourseAttributes[] activeCourses = { courses[0] };
+        CourseAttributes[] archivedCourses = { courses[1] };
+        CourseAttributes[] deletedCourses = { courses[2] };
+
+        coursesPage.verifyActiveCoursesDetails(activeCourses);
+        coursesPage.verifyArchivedCoursesDetails(archivedCourses);
+        coursesPage.verifyDeletedCoursesDetails(deletedCourses);
+
+        ______TS("verify statistics");
+        coursesPage.verifyActiveCourseStatistics(courses[0], "1", "3", "3", "1");
```
Comment (rrtheonlyone) [inline src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java:64]: This is a minor nit - when you hardcode in "1" / "3" / "3" / "1", you may run into problems when the course details are modified (you have to change two places instead of one). Perhaps getting it programmatically might be better?

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java (hunk 1 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java:180]: `verifyPresentInDatastore` is actually in `BaseTestCaseWithDatastoreAccess` which is one of the ancestors of all E2E test classes. However, I believe that the courses page test is the only test that will require the verification of archived courses in the datastore and hence I left it at that level.

File: src/e2e/java/teammates/e2e/util/BackDoor.java (hunk 2 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/util/BackDoor.java:298]: It is actually a different endpoint as archived status is not saved in the CourseAttributes but rather in InstructorAttributes. Hence the GetCourseAction endpoint cannot be used and the getCourses endpoint which gets all courses by instructor id has to be used instead.

File: src/main/java/teammates/ui/webapi/action/Action.java (hunk 3 above)
Comment (jtankw3) [reply src/main/java/teammates/ui/webapi/action/Action.java:100]: Similar to above comment, it is because archived status is saved under InstructorAttributes and in order to get the archived courses on the instructor, the request must have the instructors user id. Without doing this the user field is null when using Backdoor.

**Comment by jtankw3** [comment]
> One minor thing - when I was running them, they fail 1 out of 4 times. Was that the case for you also? (or do they always pass)

Firstly, thank you for your in depth review. 
I have looked into why the test is failing and it seems to be because 
1) The archive update in the backend takes some time and the verification is too quick - I have rectified this by adding some retries
2) The tables sometimes do not load when the page is refreshed - I could not find the root cause for this as the code for loading the tables seems to be correct. It could just be data taking too long to load from the server some times and the page displaying before the data is ready.

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    private void verifyActiveCourseStatistics(InstructorCoursesPage coursesPage, CourseAttributes course) {
+        int numSections = 0;
+        int numTeams = 0;
+        int numStudents = 0;
+        int numUnregistered = 0;
+        Set<String> sections = new HashSet<>();
+        Set<String> teams = new HashSet<>();
+
+        for (StudentAttributes student : testData.students.values()) {
+            if (student.course.equals(course.getId())) {
```
Comment (rrtheonlyone) [inline src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java:191]: Here if you want, can do:

```
if (!student.course.equals(course.getId())) {
   continue;
}
```

Then un-indent the statements below!

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+import teammates.e2e.pageobjects.InstructorCoursesPage;
+import teammates.e2e.pageobjects.InstructorHomePage;
+import teammates.e2e.util.BackDoor;
+
+/**
+ * SUT: {@link Const.WebPageURIs#INSTRUCTOR_COURSES_PAGE}.
+ */
+public class InstructorCoursesPageE2ETest extends BaseE2ETestCase {
+    CourseAttributes[] courses = new CourseAttributes[3];
+    CourseAttributes newCourse;
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java:26]: These can be `private`?

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        ______TS("verify statistics");
+        verifyActiveCourseStatistics(coursesPage, courses[0]);
+
+        ______TS("verify cannot modify without permissions");
+        coursesPage.verifyNotModifiable(courses[0].getId());
+
+        ______TS("add new course");
+        CourseAttributes[] activeCoursesWithNewCourse = { newCourse, courses[0] };
+        String[] expectedLinks = { "/web/instructor/courses/enroll?courseid=" + newCourse.getId(),
+                "/web/instructor/courses/edit?courseid=" + newCourse.getId() };
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java:75]: There are items in `Const.WebPageURIs` that can be used here.
